### PR TITLE
refactor(credentials): credential format service improvements

### PR DESCRIPTION
**Pull request by NB-MikeRichardson**
This PR covers three issues: 

#1089 (Part 1)
#1126 
#1132 

I closed the previous PR related to 1089 (as I discovered dependencies between it and #1126) so it made more sense to package them up together in one PR

### CODE REVIEW

File: packages/core/src/modules/credentials/formats/__tests__/JsonLdCredentialFormatService.test.ts [hunk 1]
```diff
@@ -149,12 +147,12 @@ const inputDoc = {
     },
     alumniOf: 'oops',
   },
-}
+})
 const verificationMethod = `8HH5gYEeNc3z7PYXmd54d4x6qAfCNrqQqEB3nS7Zfu7K#8HH5gYEeNc3z7PYXmd54d4x6qAfCNrqQqEB3nS7Zfu7K`
-const credential = JsonTransformer.fromJSON(inputDoc, W3cCredential)
+// const credential = JsonTransformer.fromJSON(inputDoc, W3cCredential)
```
Comment (TimoGlastra) [inline packages/core/src/modules/credentials/formats/__tests__/JsonLdCredentialFormatService.test.ts:152]:
```suggestion
```

File: packages/core/src/modules/credentials/formats/__tests__/JsonLdCredentialFormatService.test.ts [hunk 2]
```diff
@@ -149,12 +147,12 @@ const inputDoc = {
...
 
-const signCredentialOptions: SignCredentialOptionsRFC0593 = {
-  credential,
+const signCredentialOptions: SignCredentialOptionsRFC0593AsJson = {
+  credentialAsJson: inputDocAsJson,
```
Comment (TimoGlastra) [inline packages/core/src/modules/credentials/formats/__tests__/JsonLdCredentialFormatService.test.ts:155]: can we make the input just `credential`? 

```suggestion
  credential: inputDocAsJson,
```

File: packages/core/src/modules/credentials/formats/jsonld/JsonLdCredentialFormat.ts [hunk 3]
```diff
@@ -11,15 +11,29 @@ export interface SignCredentialOptionsRFC0593 {
...
+  credentialAsJson: JSON
+  options: JsonLdOptionsRFC0593
+}
+
+// use empty object in the acceptXXX jsonld format interface so we indicate that
+// the jsonld format service needs to be invoked
+type EmptyObject = Record<string, never>
+
+// it is an option to provide the verification method in acceptRequest
+type RequestVerificationMethod = {
```
Comment (TimoGlastra) [inline packages/core/src/modules/credentials/formats/jsonld/JsonLdCredentialFormat.ts:24]: Can we call this `JsonLdCreateRequestFormat`? This makes it consistent with the indy credential format naming

Also, can we export it and make it an interface for consistency

File: packages/core/src/modules/credentials/formats/jsonld/JsonLdCredentialFormat.ts [hunk 4]
```diff
@@ -11,15 +11,29 @@ export interface SignCredentialOptionsRFC0593 {
   options: JsonLdOptionsRFC0593
 }
 
+export interface SignCredentialOptionsRFC0593AsJson {
```
Comment (TimoGlastra) [inline packages/core/src/modules/credentials/formats/jsonld/JsonLdCredentialFormat.ts:14]: xCan we remove the AsJson part and include JsonLd in the name?

```suggestion
export interface JsonLdSignCredentialFormat {
```

File: packages/core/src/modules/credentials/formats/jsonld/JsonLdCredentialFormat.ts [hunk 5]
```diff
@@ -11,15 +11,29 @@ export interface SignCredentialOptionsRFC0593 {
...
-    createOffer: SignCredentialOptionsRFC0593
-    acceptOffer: SignCredentialOptionsRFC0593
-    createRequest: SignCredentialOptionsRFC0593
-    acceptRequest: JsonLdAcceptRequestOptions
+    createProposal: SignCredentialOptionsRFC0593AsJson
+    acceptProposal: EmptyObject
+    createOffer: SignCredentialOptionsRFC0593AsJson
+    acceptOffer: EmptyObject
+    createRequest: SignCredentialOptionsRFC0593AsJson
+    acceptRequest: RequestVerificationMethod
```
Comment (TimoGlastra) [inline packages/core/src/modules/credentials/formats/jsonld/JsonLdCredentialFormat.ts:37]: This is missing typing for the `formatData`. See indy credential format as example

File: packages/core/src/modules/credentials/formats/jsonld/JsonLdCredentialFormatService.ts [hunk 6]
```diff
@@ -146,9 +141,7 @@ export class JsonLdCredentialFormatService extends CredentialFormatService<JsonL
       throw new AriesFrameworkError('Missing jsonld payload in createOffer')
     }
 
-    const jsonLdCredential = new JsonLdCredentialDetail(jsonLdFormat)
-
-    MessageValidator.validateSync(jsonLdCredential)
+    MessageValidator.validateSync(jsonLdFormat.credentialAsJson)
```
Comment (TimoGlastra) [inline packages/core/src/modules/credentials/formats/jsonld/JsonLdCredentialFormatService.ts:144]: Mike, made this comment multiple times already, You can't validate JSON. It needs to be a class instance!!

We can parse it as an `JsonLdCredentialDetail` and then validate....

**Review by TimoGlastra — CHANGES_REQUESTED** [review]

File: packages/core/src/modules/credentials/formats/__tests__/JsonLdCredentialFormatService.test.ts (hunk 2 above)
Comment (NB-MikeRichardson) [reply packages/core/src/modules/credentials/formats/__tests__/JsonLdCredentialFormatService.test.ts:155]: Yes, this was really to aid in development, as it is nice to be able to see the new field distinguished from the old one in certain places.

File: packages/core/src/modules/credentials/formats/jsonld/JsonLdCredentialFormat.ts (hunk 5 above)
Comment (NB-MikeRichardson) [reply packages/core/src/modules/credentials/formats/jsonld/JsonLdCredentialFormat.ts]: Not entirely sure what this is used for, but I've added it anyway.

**Comment by NB-MikeRichardson** [comment]
I have resolved all conversations relating to this PR: it can now be re-reviewed

File: packages/core/src/modules/credentials/formats/jsonld/JsonLdCredentialFormat.ts [hunk 7]
```diff
@@ -1,25 +1,48 @@
...
-export interface JsonLdAcceptRequestOptions extends SignCredentialOptionsRFC0593 {
-  verificationMethod?: string
+export interface JsonCredential {
+  '@context': Array<string> | JsonObject
+  id?: string
+  type: Array<string>
+  issuer: string | IssuerOptions
+  issuanceDate: string
+  expirationDate?: string
+  credentialSubject: SingleOrArray<JsonObject>
```
Comment (TimoGlastra) [inline packages/core/src/modules/credentials/formats/jsonld/JsonLdCredentialFormat.ts:15]: this will allow for extra properties to be defined on the credential

```suggestion
  credentialSubject: SingleOrArray<JsonObject>
  [key: string]: unknown
```

File: packages/core/src/modules/credentials/formats/jsonld/JsonLdCredentialFormat.ts [hunk 8]
```diff
@@ -1,25 +1,48 @@
...
+    createOffer: JsonLdSignCredentialFormat
+    acceptOffer: EmptyObject
+    createRequest: JsonLdSignCredentialFormat
+    acceptRequest: JsonLdCreateRequestFormat
+  }
+  formatData: {
+    proposal: JsonLdSignCredentialFormat
+    offer: JsonLdSignCredentialFormat
+    request: JsonLdSignCredentialFormat
+    credential: W3cCredential
```
Comment (TimoGlastra) [inline packages/core/src/modules/credentials/formats/jsonld/JsonLdCredentialFormat.ts:46]: This will be a JSON representation of the credential, not an instance of W3cCredential.

I think we should create a new `JsonVerifiableCredential`  that extends `JsonCredential` and adds the proof property:

```
export interface JsonVerifiableCredential extends JsonCredential {
  proof: SingleOrArray<{
    type: string
    proofPurpose: string
    verificationMethod: string
    created: string
	  domain?: string
	  challenge?: string
	  jws?: string
	  proofValue?: string
	  nonce?: string
[key: string]: unknown
  }>
}
```

File: packages/core/src/modules/credentials/formats/jsonld/JsonLdCredentialFormatService.ts [hunk 9]
```diff
@@ -86,38 +84,30 @@ export class JsonLdCredentialFormatService extends CredentialFormatService<JsonL
    * @param options the options needed to accept the proposal
    */
   public async processProposal(agentContext: AgentContext, { attachment }: FormatProcessOptions): Promise<void> {
-    const credProposalJson = attachment.getDataAsJson<SignCredentialOptionsRFC0593>()
+    const credProposalJson = attachment.getDataAsJson<JsonLdSignCredentialFormat>()
```
Comment (TimoGlastra) [inline packages/core/src/modules/credentials/formats/jsonld/JsonLdCredentialFormatService.ts:87]: We should be careful here. The `JsonLdSignCredentialFormat` is for the input (the API) we use in AFJ. Although this is currently the same as the RFC, this doesn't necesarily stay the same. I think we shouldn't reuse interfaces for different purposes like this, as it increases the chance of having invalid interfaces in some places of the codebase.

File: packages/core/src/modules/credentials/formats/jsonld/JsonLdCredentialFormatService.ts [hunk 10]
```diff
@@ -207,42 +193,35 @@ export class JsonLdCredentialFormatService extends CredentialFormatService<JsonL
       throw new AriesFrameworkError('Missing jsonld payload in createRequest')
     }
 
-    const jsonLdCredential = new JsonLdCredentialDetail(jsonLdFormat)
-    MessageValidator.validateSync(jsonLdCredential)
+    MessageValidator.validateSync(jsonLdFormat.credential)
```
Comment (TimoGlastra) [inline packages/core/src/modules/credentials/formats/jsonld/JsonLdCredentialFormatService.ts:196]: Can't validate json. Need to transform it

File: packages/core/src/modules/credentials/formats/jsonld/JsonLdCredentialFormatService.ts [hunk 11]
```diff
@@ -317,20 +300,17 @@ export class JsonLdCredentialFormatService extends CredentialFormatService<JsonL
...
       throw new AriesFrameworkError('Credential arrays are not supported')
     } else {
       // do checks here
-      this.compareCredentialSubject(credential, request.credential)
-      this.compareProofs(credential.proof, request.options)
+      this.compareCredentialSubject(credential, requestAsJson)
+      this.compareProofs(credential.proof, requestAsJson)
     }
 
     const verifiableCredential = await this.w3cCredentialService.storeCredential(agentContext, {
```
Comment (TimoGlastra) [inline packages/core/src/modules/credentials/formats/jsonld/JsonLdCredentialFormatService.ts:316]: This doesn't yet verify the credential as described in #1132 (the PR body mentions it fixes this issue).

File: packages/core/src/modules/credentials/protocol/v2/__tests__/v2.ldproof.credentials.propose-offerED25519.test.ts [hunk 12]
```diff
@@ -325,26 +316,23 @@ describe('credentials', () => {
...
           'https://www.w3.org/2018/credentials/v1',
           'https://w3id.org/citizenship/v1',
           'https://w3id.org/security/bbs/v1',
         ],
         id: 'https://issuer.oidp.uscis.gov/credentials/83627465',
         type: ['VerifiableCredential', 'PermanentResidentCard'],
         issuer: 'did:key:z6Mkgg342Ycpuk263R9d8Aq6MUaxPn1DDeHyGo38EefXmgDL',
-        identifier: '83627465',
-        name: 'Permanent Resident Card',
-        description: 'Government of Example Permanent Resident Card.',
```
Comment (TimoGlastra) [inline packages/core/src/modules/credentials/protocol/v2/__tests__/v2.ldproof.credentials.propose-offerED25519.test.ts:342]: With the type suggestion I made we can keep these properties in

**Review by TimoGlastra — CHANGES_REQUESTED** [review]

File: packages/core/src/modules/credentials/formats/jsonld/JsonLdCredentialFormatService.ts (hunk 9 above)
Comment (NB-MikeRichardson) [reply packages/core/src/modules/credentials/formats/jsonld/JsonLdCredentialFormatService.ts:87]: I have put back the SignCredentialOptionsRFC0593 for use internally (only) so if there is a divergence we are covered.

I assume that is what we want to do here, please let me know if there is a better way of fixing this.

File: packages/core/src/modules/credentials/formats/jsonld/JsonLdCredentialFormatService.ts (hunk 11 above)
Comment (NB-MikeRichardson) [reply packages/core/src/modules/credentials/formats/jsonld/JsonLdCredentialFormatService.ts:316]: I assume this should be a call to the w3cCredentialService `verifiyCredential()` method.

**Comment by NB-MikeRichardson** [comment]
All issues resolved, but in adding the verifyCredential call() some of the tests now take too long and time out. I have (temporarily) upped the timeout to 200s. I understand a better solution to be setting the custom document loader in the w3cCredentialService, but I am not sure how to do this from the level of the e2e tests.

See v2/__tests__v2.ldproof.credentials-auto-accept.test.ts

**Comment by TimoGlastra** [comment]
Will https://github.com/hyperledger/aries-framework-javascript/issues/1127 be addressed in a separate PR @NB-MikeRichardson?

File: packages/core/src/modules/credentials/protocol/v2/__tests__/v2.ldproof.credentials.propose-offerED25519.test.ts [hunk 13]
```diff
@@ -325,30 +319,30 @@ describe('credentials', () => {
...
+        '@context': [
           'https://www.w3.org/2018/credentials/v1',
           'https://w3id.org/citizenship/v1',
           'https://w3id.org/security/bbs/v1',
         ],
         id: 'https://issuer.oidp.uscis.gov/credentials/83627465',
         type: ['VerifiableCredential', 'PermanentResidentCard'],
         issuer: 'did:key:z6Mkgg342Ycpuk263R9d8Aq6MUaxPn1DDeHyGo38EefXmgDL',
-        identifier: '83627465',
-        name: 'Permanent Resident Card',
```
Comment (TimoGlastra) [inline packages/core/src/modules/credentials/protocol/v2/__tests__/v2.ldproof.credentials.propose-offerED25519.test.ts:341]: Any reason to move these to the credentialSubject? The example has them top level. Does TS give errors in this case? We should resolved that then

File: packages/core/src/modules/credentials/protocol/v2/__tests__/v2.ldproof.credentials.propose-offerED25519.test.ts [hunk 14]
```diff
@@ -33,9 +32,6 @@ describe('credentials', () => {
     id: 'https://issuer.oidp.uscis.gov/credentials/83627465',
     type: ['VerifiableCredential', 'PermanentResidentCard'],
     issuer: 'did:key:z6Mkgg342Ycpuk263R9d8Aq6MUaxPn1DDeHyGo38EefXmgDL',
-    identifier: '83627465',
-    name: 'Permanent Resident Card',
-    description: 'Government of Example Permanent Resident Card.',
```
Comment (TimoGlastra) [inline packages/core/src/modules/credentials/protocol/v2/__tests__/v2.ldproof.credentials.propose-offerED25519.test.ts:38]: Same here, is this change needed because of TS errors?

File: packages/core/src/modules/credentials/formats/jsonld/JsonLdCredentialFormat.ts [hunk 15]
```diff
@@ -1,25 +1,61 @@
...
 
+// use this interface internally as the above may diverge in future
 export interface SignCredentialOptionsRFC0593 {
-  credential: W3cCredential
+  credential: JsonCredential
   options: JsonLdOptionsRFC0593
 }
 
+export interface JsonVerifiableCredential extends JsonLdSignCredentialFormat {
+  proof: LinkedDataProof
```
Comment (TimoGlastra) [inline packages/core/src/modules/credentials/formats/jsonld/JsonLdCredentialFormat.ts:32]: LinkedDataProof is a class, so it won't have the right interface. We should create a custom interface that is just for the json representation of the jsonld credential. See my comment here of what the interface could look like: https://github.com/hyperledger/aries-framework-javascript/pull/1140#discussion_r1045566257

**Review by TimoGlastra — COMMENTED** [review]
Mostly looks good @NB-MikeRichardson, some final remarks. I'm waiting for #1159 to be merged then we can reduce the timeout and we should be good to merge 👍

File: packages/core/src/modules/credentials/formats/jsonld/JsonLdCredentialFormat.ts (hunk 8 above)
Comment (NB-MikeRichardson) [reply packages/core/src/modules/credentials/formats/jsonld/JsonLdCredentialFormat.ts]: Took out LinkedDataProof, replaced with above interface

**Comment by NB-MikeRichardson** [comment]
Comments resolved, just waiting for the custom document loader PR to be merged and then I think we are ready to merge this one as well. Although no tests are timing out in this PR (the culprit is in the PPV2 PR). That said we should include the custom loader here.

**Comment by NB-MikeRichardson** [comment]
> @NB-MikeRichardson if you can remove the timemoutMs changes and merge with main, we can get this merged

Done. Merged with the new custom loader code, timeout removed, test passes. Code committed...

Good to go.
